### PR TITLE
Adding bootstrap endpoint for App and variants

### PR DESCRIPTION
**Pull request by matzew**
https://issues.jboss.org/browse/AGPUSH-1469

One (multipart) endpoint to create Windows, Android and/or iOS variant, along the push application

### CODE REVIEW

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+public class BootstrapEndpoint extends AbstractBaseEndpoint {
+
+    @Inject
+    private PushApplicationService pushAppService;
+
+    @Inject
+    protected GenericVariantService variantService;
+
+    /**
+     * Convenience endpoint to create a complete Push Application with and given variants
```
Comment (sebastienblanc) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java]: `with and given variants` is not so clear,  `with a set of variants` ?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+     *
+     * @param form              Form containing data for Push App and variants
+     * @return                  created {@link PushApplication}
+     *
+     * @statuscode 201 The PushApplication Variant created successfully
```
Comment (sebastienblanc) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java]: `The PushApplication Variant` => `The PushApplication with Variants`

File: pom.xml [hunk 3]
```diff
@@ -168,8 +168,8 @@
             Options to override the compiler arguments directly on the compiler arument line to separate between what
             the IDE understands as the source level and what the Maven compiler actually use.
         -->
-        <maven.compiler.argument.target>${maven.compiler.target}</maven.compiler.argument.target>
-        <maven.compiler.argument.source>${maven.compiler.source}</maven.compiler.argument.source>
+        <maven.compiler.argument.target>1.7</maven.compiler.argument.target>
+        <maven.compiler.argument.source>1.7</maven.compiler.argument.source>
```
Comment (sebastienblanc) [inline pom.xml]: is that needed for the PR ?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        return Response.created(
+                UriBuilder.fromResource(PushApplicationEndpoint.class)
+                        .path(String.valueOf(pushApplication.getPushApplicationID()))
+                        .build())
+                .entity(pushApplication)
+                .build();
+    }
+}
```
Comment (sebastienblanc) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java]: What happens if the creation of one variant fails ? Is it rolled back ?

File: pom.xml (hunk 3 above)
Comment (matzew) [inline pom.xml]: yes - switch (w/ string)

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java (hunk 4 above)
Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java]: it will create just the other variants

Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/applications/BootstrapEndpoint.java]: but there is valiation up front, so good coverage there

**Comment by lfryc** [comment]
We discussed this with @sebastienblanc now and I had this idea:

In order to reuse the logic from the UPS UI (specifically "Add Variant" modal panel) for bootstrapping app with variants (it can be later adopted by NGUI during when "Enable Push" is performed),

it would be better if a request for this endpoint look like:

````
{
    pushApplicationName: ...,
    variants: [
        { type: "android", variantName: ..., googleKey: ..., projectNumber: ... },
        { type: "ios", .... }
    ]
}
````

**Comment by lfryc** [comment]
Ok, discussed over hangout, we are not able to simply follow dot notation with application/www-form-multipart, so let's use this notation.

**Pull request merged**